plumbing/format/packfile/cache.go

```diff
@@ -0,0 +1,46 @@
+package packfile
+
+import (
+	"gopkg.in/src-d/go-git.v4/plumbing"
+)
+
+type cache struct {
+	objects     map[plumbing.Hash]plumbing.EncodedObject
+	maxElements int
+	order       []plumbing.Hash
+}
+
+func newCache() *cache {
+	return &cache{
+		objects:     make(map[plumbing.Hash]plumbing.EncodedObject),
+		maxElements: 1,
+		order:       make([]plumbing.Hash, 1),
+	}
+}
+
+func (c *cache) SetMaxElements(max int) {
+	if max <= 0 {
+		max = 1
+	}
+	c.maxElements = max
+}
+
+func (c *cache) Add(o plumbing.EncodedObject) {
+	if len(c.order) >= c.maxElements {
+		d, order := c.order[0], c.order[1:]
```

Comment:
This is quite inefficient, we should probably implement a queue as a circular list backed by a slice.